lib/chef-dk/policyfile/undo_record.rb

```diff
@@ -0,0 +1,135 @@
+          self.policy_group = data["policy_group"]
+          self.data = data["data"]
+          self
+        end
+
+        def for_serialization
+          {
+            "policy_name" => policy_name,
+            "policy_group" => policy_group,
+            "data" => data
+          }
+        end
+
+      end
+
+      attr_reader :policy_groups
+
+      attr_reader :policy_revisions
+
+      def initialize
+        reset!
+      end
+
+      def ==(other)
+        other.kind_of?(UndoRecord) &&
+          other.policy_groups == policy_groups &&
+          other.policy_revisions == policy_revisions
+      end
+
+      def add_policy_group(name)
+        @policy_groups << name
+      end
+
+      def add_policy_revision(policy_name, policy_group, data)
+        @policy_revisions << PolicyGroupRestoreData.new(policy_name, policy_group, data)
+      end
+
+      def load(data)
+        reset!
+
+        unless data.kind_of?(Hash)
+          raise InvalidUndoRecord, "Undo data is incorrectly formatted. Must be a Hash, got '#{data}'."
+        end
+        missing_fields = %w[ format_version backup_data ].select { |key| !data.key?(key) }
+        unless missing_fields.empty?
+          raise InvalidUndoRecord, "Undo data is missing mandatory field(s) #{missing_fields.join(', ')}. Undo data: '#{data}'"
+        end
+        policy_data = data["backup_data"]
+        unless policy_data.kind_of?(Hash)
+          raise InvalidUndoRecord, "'backup_data' in the undo record is incorrectly formatted. Must be a Hash, got '#{policy_data}'"
+        end
+        missing_policy_data_fields = %w[ policy_groups policy_revisions ].select { |key| !policy_data.key?(key) }
+        unless missing_policy_data_fields.empty?
+          raise InvalidUndoRecord,
+            "'backup_data' in the undo record is missing mandatory field(s) #{missing_policy_data_fields.join(', ')}. Backup data: #{policy_data}"
+        end
+
+        policy_groups = policy_data["policy_groups"]
+
+        unless policy_groups.kind_of?(Array)
+          raise InvalidUndoRecord,
+            "'policy_groups' data in the undo record is incorrectly formatted. Must be an Array, got '#{policy_groups}'"
+        end
+
+        @policy_groups = policy_groups
+
+        policy_revisions = policy_data["policy_revisions"]
+        unless policy_revisions.kind_of?(Array)
+          raise InvalidUndoRecord,
+            "'policy_revisions' data in the undo record is incorrectly formatted. Must be an Array, got '#{policy_revisions}'"
+        end
+
+        policy_revisions.each do |revision|
+          unless revision.kind_of?(Hash)
+            raise InvalidUndoRecord,
+              "Invalid item in 'policy_revisions' in the undo record. Must be a Hash, got '#{revision}'"
+          end
+
+          @policy_revisions << PolicyGroupRestoreData.new.load(revision)
+        end
+
+        self
+      end
+
+      def for_serialization
+        {
+          "format_version" => 1,
```

Comment:
This is here so we can version the way the records are formatted, i.e., we could introduce a version 2 in the future and then use `format_version` when reading to pick which code branch we need to read the data. But right now there's only version 1.